fdbserver/worker.actor.cpp

```diff
@@ -592,6 +592,150 @@ ACTOR Future<Void> registrationClient(Reference<AsyncVar<Optional<ClusterControl
+ACTOR Future<Void> healthMonitor(Reference<AsyncVar<Optional<ClusterControllerFullInterface>>> ccInterface,
+                                 WorkerInterface interf,
+                                 LocalityData locality,
+                                 Reference<AsyncVar<ServerDBInfo>> dbInfo) {
+	loop {
+		Future<Void> nextHealthCheckDelay = Never();
+		if (dbInfo->get().recoveryState >= RecoveryState::ACCEPTING_COMMITS &&
+		    addressInDbAndPrimaryDc(interf.addresses(), dbInfo) && ccInterface->get().present()) {
+			nextHealthCheckDelay = delay(SERVER_KNOBS->WORKER_HEALTH_MONITOR_INTERVAL);
+			auto* allPeers = FlowTransport::transport().getAllPeers();
+			for (auto& [address, peer] : *allPeers) {
+				if (peer->pingLatencies.getPopulationSize() < 50) {
+					// Ignore peers that do not have enough samples.
+					// TODO(zhewu): Currently, FlowTransport latency monitor clears ping latency samples on a regular
+					//              basis, which may affect the measurement count. Currently,
+					//              WORKER_HEALTH_MONITOR_INTERVAL is much smaller than the ping clearance interval, so
+					//              it may be ok. If this ends to be a problem, we need to consider keep track of last
+					//              ping latencies logged.
+					continue;
+				}
+
+				if (!addressInDbAndPrimaryDc(address, dbInfo)) {
+					// Ignore the servers that are not in the database's transaction system and not in the primary DC.
+					// Note that currently we are not monitor storage servers, since lagging in storage servers today
+					// already can trigger server exclusion by data distributor.
+					continue;
+				}
+
+				if (peer->pingLatencies.percentile(SERVER_KNOBS->PEER_LATENCY_DEGRADATION_PERCENTILE) >
+				        SERVER_KNOBS->PEER_LATENCY_DEGRADATION_THRESHOLD ||
+				    peer->timeoutCount / (double)(peer->pingLatencies.getPopulationSize()) >
+				        SERVER_KNOBS->PEER_TIMEOUT_PERCENTAGE_DEGRADATION_THRESHOLD) {
+					// This is a degraded peer.
+					TraceEvent("HealthMonitorDetectDegradedPeer")
+					    .detail("Peer", address)
+					    .detail("Elapsed", now() - peer->lastLoggedTime)
+					    .detail("MinLatency", peer->pingLatencies.min())
+					    .detail("MaxLatency", peer->pingLatencies.max())
+					    .detail("MeanLatency", peer->pingLatencies.mean())
+					    .detail("MedianLatency", peer->pingLatencies.median())
+					    .detail("CheckedPercentile", SERVER_KNOBS->PEER_LATENCY_DEGRADATION_PERCENTILE)
+					    .detail("CheckedPercentileLatency",
+					            peer->pingLatencies.percentile(SERVER_KNOBS->PEER_LATENCY_DEGRADATION_PERCENTILE))
+					    .detail("Count", peer->pingLatencies.getPopulationSize())
+					    .detail("TimeoutCount", peer->timeoutCount);
+
+					// TODO(zhewu): Keep track of degraded peers and send them to cluster controller.
+				}
+			}
+		}
+		choose {
+			when(wait(nextHealthCheckDelay)) {}
+			when(wait(ccInterface->onChange())) {}
+			when(wait(dbInfo->onChange())) {}
```

Comment:
If `dbInfo` changed a few times (e.g., during recovery and after accepting commits), then the above traceevent may be generated more frequently than the `WORKER_HEALTH_MONITOR_INTERVAL` interval, which is probably not desirable. An easy fix can be adding a `suppress` to the event.

Reply:
Good point, done.